proposals/pattern-based-fixed.md

```diff
@@ -0,0 +1,133 @@
+# Pattern-based `fixed` statement
+
+* [x] Proposed
+* [ ] Prototype: Not Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Not Started
+
+## Summary
+[summary]: #summary
+
+Introduce a pattern that would allow types to participate in `fixed` statements. 
+
+## Motivation
+[motivation]: #motivation
+
+The language provides a mechanism for pinning managed data and obtain a native pointer to the underlying buffer.
+
+```C#
+fixed(byte* ptr = byteArray)
+{
+   // ptr is a native pointer to the first element of the array
+   // byteArray is protected from being moved/collected by the GC for the duration of this block 
+}
+
+```
+
+The set of types that can participate in `foreach` is hardcoded and limited to arrays and `System.String`. Hardcoding "special" types does not scale when new primitives such as `ImmutableArray<T>`, `Span<T>`, `Utf8String` are introduced. 
+
+In addition, the current solution for `System.String` relies on a fairly rigid API. The shape of the API implies that `System.String` is a contiguous object that embeds UTF16 encoded data at a fixed offset from the object header. Such approach has been found problematic in several proposals that could require changes to the underlying layout. 
+It would be desirable to be able to switch to something more flexible that decouples `System.String` object from its internal representation for the purpose of unmanaged interop. 
+
+## Detailed design
+[design]: #detailed-design
+
+## *Pattern* ##
+A viable pattern-based �fixed� need to:
+-	Provide the managed references to pin the instance and to initialize the pointer (preferably this is the same reference)
+-	Convey unambiguously the type of the unmanaged element   (i.e. �char� for �string�)
+-	Prescribe the behavior in "empty" case when there is nothing to refer to. 
+-	Should not push API authors toward design decisions that hurt the use of the type outside of `fixed`.
+
+
+I think the above could be satisfied by recognizing a specially named ref-returning member:
+ `ref [readonly] T DangerousGetPinnableReference()`.
+
+In order to be used by the `foreach` statement the following conditions must be met:
+
+1)	There is only one such member provided for a type.
```

Comment:
Extension methods would be supported, presumably?

Reply:
I was thinking about this and was not sure. The method would typically be very coupled with internal implementation of the type. 

On the other hand if type cannot be changed, can extension be provided? (likely encapsulating unsavory hacks against implementation, but at least in one place)

Extensions can be allowed, but not everything that can be allowed should be allowed. I will add this to the list of things to discuss.

Reply:
> The method would typically be very coupled with internal implementation of the type.

Yes, but that doesn't prevent it from being an extension method, which might be done purely for usability purposes.  For example, today these DangerousGetPinnableReference methods on span/readonlyspan are marked as non-browsable, so that they don't show up in normal IntelliSense browsing.  But that's a bad experience when you actually do need them, as it's a long name for which you don't get completion support.  We're talking instead about making them extension methods that live on the System.Runtime.InteropServices.MemoryMarshal class, also in Corelib, without the attribute.  So such DangerousGetPinnableReference methods could be extensions but still live in the same assembly.

Reply:
> We're talking instead about making them non-extension methods that live on the System.Runtime.InteropServices.MemoryMarshal class, also in Corelib, without the attribute.

We should have both the extension method for direct use, and (instance?) method for the pinning pattern.

- The (instance?) method for pinning should have the safe-guard for empty spans to return null.
- The extension method on MemoryMarshal should not have the safe-guard for empty spans.

Reply:
> and (instance?) method for the pinning pattern

Right, we just planned to defer the pinning pattern one until it was actually a language feature.

> We should have both the extension method for direct use, 

But not an extension method, right?  Just a static method on a marshaling class.

Reply:
> But not an extension method, right

Right.